### ISSUE TITLE
Can't follow from GoToSocial or Takahe

### ISSUE THREAD

**Issue by kvibber** [writefreely/writefreely#661]
I'm able to follow a WriteFreely blog from Mastodon, but I haven't been able to get Takahe or GoToSocial to follow successfully. `writefreely --debug` doesn't tell me anything, but I do get the following in the nginx log:

Follow request from GoToSocial:

```
45.79.78.63 - - [16/Mar/2023:01:41:30 +0000] "POST /api/collections/KelsonWrites/inbox HTTP/1.1" 301 162 "-" "gotosocial (+https://gts.keysmash.xyz) gotosocial/0.7.1 git-adb5966" "-"
45.79.78.63 - - [16/Mar/2023:01:41:30 +0000] "GET /api/collections/KelsonWrites/inbox HTTP/1.1" 404 19 "http://log.kvibber.com/api/collections/KelsonWrites/inbox" "gotosocial (+https://gts.keysmash.xyz) gotosocial/0.7.1 git-adb5966" "-"
```

Follow request from Takahe:

```
35.209.52.146 - - [16/Mar/2023:01:43:26 +0000] "GET /api/collections/KelsonWrites HTTP/1.1" 301 162 "-" "python-httpx/0.23.3 (Takahe/0.8.0; +https://takahe.social/)" "-"
35.209.52.146 - - [16/Mar/2023:01:43:26 +0000] "GET /api/collections/KelsonWrites HTTP/1.1" 302 0 "-" "python-httpx/0.23.3 (Takahe/0.8.0; +https://takahe.social/)" "-"
35.209.52.146 - - [16/Mar/2023:01:43:26 +0000] "GET / HTTP/1.1" 200 3724 "-" "python-httpx/0.23.3 (Takahe/0.8.0; +https://takahe.social/)" "-"
```

**Comment by thebaer** [comment]
Thanks for bringing this up. Are you sure the application logs don't show anything when running with the `-debug` flag? You should see full HTTP dumps on those requests when running WF that way.

**Comment by gaviscapes** [comment]
Following up on this, my logs when following from gotosocial show this

```
2023/10/09 05:12:09 "POST /api/collections/gavi/inbox" 200 103.402547ms "gotosocial (+https://sleepy.ink) gotosocial/0.11.1 git-c7a46e0"
2023/10/09 05:12:11 POST
ERROR: 2023/10/09 05:12:11 activitypub.go:442: Unable to make activity POST: Post "": unsupported protocol scheme ""

```

**Comment by nullium21** [comment]
was trying to follow from glitch-soc (mastodon fork) and got the same errors:
```
2024/02/14 14:02:04 Fetching actor https://tech.lgbt/users/lina locally
2024/02/14 14:02:04 Not found; fetching actor https://tech.lgbt/users/lina remotely
2024/02/14 14:02:04 GET https://tech.lgbt/users/lina
2024/02/14 14:02:04 "GET /api/collections/blog.lina.moe" 200 378.563µs "http.rb/5.1.1 (Mastodon/4.3.0-alpha.1+glitch; +https://tech.lgbt/)"
2024/02/14 14:02:05 "GET /api/collections/lina" 200 304.482µs "http.rb/5.1.1 (Mastodon/4.3.0-alpha.1+glitch; +https://tech.lgbt/)"
2024/02/14 14:02:05 "GET /api/collections/lina" 200 299.002µs "http.rb/5.1.1 (Mastodon/4.3.0-alpha.1+glitch; +https://tech.lgbt/)"
2024/02/14 14:02:05 "POST /api/collections/lina/inbox" 200 1.30685025s "http.rb/5.1.1 (Mastodon/4.3.0-alpha.1+glitch; +https://tech.lgbt/)"
2024/02/14 14:02:07 POST 
ERROR: 2024/02/14 14:02:07 activitypub.go:456: Unable to make activity POST: Post "": unsupported protocol scheme ""
```

the mastodon instance i'm using (https://tech.lgbt) has "authorized fetch" enabled, making it require signatures on all requests if i remember correctly, and this message captured with `-debug` seems to suggest that's the reason:
```
2024/02/14 14:04:52 Status  : 401 Unauthorized
2024/02/14 14:04:52 Response: {"error":"Verification failed for lina@blog.lina.moe https://blog.lina.moe/api/collections/lina using rsa-sha256 (RSASSA-PKCS1-v1_5 with SHA-256)","signed_string":"(request-target): get /users/lina\ndate: Wed, 14 Feb 2024 14:04:50 GMT\nhost: tech.lgbt\ndigest: SHA-256=47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=","signature":"L3KE6RbLSV8eUJ8zSpTGsXkI6tlDLfm0mLFwChiJ1s2wmCqxLgwcqOahuq2cEJfuKMSs30dyaDBXQ0T9ZkSeLQhopu8BmJV+DR4uAMLhploutQ6fFXqg1lVzw3QuJpwLfgb/wi14RoWCmul/sBT4X87ncfHJ4rhW6F8pauJMqjpckXzZy4IiqSipKbxc7KYGs4BJsi/NhNsshJifNBpBrgTxFUeOyeShYx3luqI/VxoBEnDCyPBgO4oerRL9RekhgKblpwfpQNS3vgSttbLl3wtAJAvJ4RQj81el/YlLy65b5tB6w6lw1EC2S8JfSKFHXBDnrT41rLXmAtruiAUaIA=="}
```

**Comment by gytisrepecka** [comment]
Apparently following WriteFreely blog from Mastodon involves some extra communication which I don't observe when GoToSocial is attempting to follow WriteFreely blog.

Here is debug information from my exercise:

* Source 1: user which attempts to follow from GoToSocial: `@gytisrepecka@social.gyt.is`
* Source 2: user which attempts to follow from Mastodon: `@inretio@fedi.lt`
* Target: user which is being followed on WriteFreely: `@gytis@fedi.dev`

Debug log from target (with hashes and IPs hidden) - WriteFreely:

```log
2024/02/18 15:13:05 Rec'd! "POST /api/collections/gytis/inbox HTTP/1.0\r\nHost: fedi.dev\r\nConnection: close\r\nAccept-Charset: utf-8\r\nAccept-Encoding: gzip\r\nConnection: close\r\nContent-Length: 297\r\nContent-Type: application/ld+json; profile=\"https://www.w3.org/ns/activitystreams\"\r\nDate: Sun, 18 Feb 2024 13:13:05 GMT\r\nDigest: SHA-256=HASH_HERE\r\nSignature: keyId=\"https://social.gyt.is/users/gytisrepecka/main-key\",algorithm=\"hs2019\",headers=\"(request-target) host date digest\",signature=\"SIGNATURE_HERE\"\r\nUser-Agent: gotosocial (+https://social.gyt.is) gotosocial/0.13.3 git-fb3e3ca\r\nX-Forwarded-For: IP_HERE\r\nX-Real-Ip: IP_HERE\r\n\r\n{\"@context\":\"https://www.w3.org/ns/activitystreams\",\"actor\":\"https://social.gyt.is/users/gytisrepecka\",\"id\":\"https://social.gyt.is/users/gytisrepecka/follow/FOLLOW_ID_HERE\",\"object\":\"https://fedi.dev/api/collections/gytis\",\"to\":\"https://fedi.dev/api/collections/gytis\",\"type\":\"Follow\"}"
2024/02/18 15:13:05 Follow: {"@context":["https://www.w3.org/ns/activitystreams"],"actor":"https://social.gyt.is/users/gytisrepecka","id":"https://social.gyt.is/users/gytisrepecka/follow/FOLLOW_ID_HERE","object":"https://fedi.dev/api/collections/gytis","to":"https://fedi.dev/api/collections/gytis","type":"Follow"}

RIGHT HERE Mastodon does additional communication which GoToSocial doesn't!

2024/02/18 15:13:05 Fetching actor https://social.gyt.is/users/gytisrepecka locally
2024/02/18 15:13:05 "POST /api/collections/gytis/inbox" 200 3.146448ms "gotosocial (+https://social.gyt.is) gotosocial/0.13.3 git-fb3e3ca"
2024/02/18 15:13:07 POST 
ERROR: 2024/02/18 15:13:07 activitypub.go:556: Can't dump: unsupported protocol scheme ""
ERROR: 2024/02/18 15:13:07 activitypub.go:456: Unable to make activity POST: Post "": unsupported protocol scheme ""
```

That extra communication happening between Mastodon and WriteFreely is following:

```log
2024/02/18 15:25:48 "POST /api/collections/gytis/inbox" 200 4.035734ms "http.rb/5.1.1 (Mastodon/4.2.7; +https://fedi.lt/)"
2024/02/18 15:25:50 POST https://fedi.lt/users/inretio/inbox
2024/02/18 15:25:50 POST /users/inretio/inbox HTTP/1.1
Host: fedi.lt
User-Agent: Go (WriteFreely/0.15.0; +https://fedi.dev)
Content-Length: 379
Content-Type: application/activity+json
Date: Sun, 18 Feb 2024 13:25:50 GMT
Digest: SHA-256=KqJl03Lln1N4wllcjY3Z7jCrYS/qj6WnlwqIpXJHSI4=
Signature: keyId="https://fedi.dev/api/collections/gytis#main-key",algorithm="rsa-sha256",headers="(request-target) date host digest",signature="SIGNATURE_HERE"
Accept-Encoding: gzip
{"@context":["https://www.w3.org/ns/activitystreams"],"actor":null,"object":{"actor":"https://fedi.lt/users/inretio","id":"https://fedi.lt/users/inretio#follows/915/undo","object":{"actor":"https://fedi.lt/users/inretio","id":"https://fedi.lt/ID_HERE","object":"https://fedi.dev/api/collections/gytis","type":"Follow"},"type":"Undo"},"type":"Accept"}
2024/02/18 15:25:50 Status  : 202 Accepted
2024/02/18 15:26:00 Rec'd! "POST /api/collections/gytis/inbox HTTP/1.0\r\nHost: fedi.dev\r\nConnection: close\r\nAccept-Encoding: gzip\r\nConnection: close\r\nContent-Length: 218\r\nContent-Type: application/activity+json\r\nDate: Sun, 18 Feb 2024 13:26:00 GMT\r\nDigest: SHA-256=HASH_HERE\r\nSignature: keyId=\"https://fedi.lt/users/inretio#main-key\",algorithm=\"rsa-sha256\",headers=\"(request-target) host date digest content-type\",signature=\"SIGNATURE_HERE\"\r\nUser-Agent: http.rb/5.1.1 (Mastodon/4.2.7; +https://fedi.lt/)\r\nX-Forwarded-For: IP_HERE\r\nX-Real-Ip: IP_HERE\r\n\r\n{\"@context\":\"https://www.w3.org/ns/activitystreams\",\"id\":\"https://fedi.lt/ID_HERE\",\"type\":\"Follow\",\"actor\":\"https://fedi.lt/users/inretio\",\"object\":\"https://fedi.dev/api/collections/gytis\"}"
2024/02/18 15:26:00 Follow: {"@context":["https://www.w3.org/ns/activitystreams"],"actor":"https://fedi.lt/users/inretio","id":"https://fedi.lt/ID_HERE","object":"https://fedi.dev/api/collections/gytis","type":"Follow"}
```

I have submitted [an issue to GoToSocial](https://github.com/superseriousbusiness/gotosocial/issues/2658) since there seems to be those extra steps that Mastodon does which GoToSocial doesn't. But would be good to receive feedback from both sides 🤓 Thanks!

**Comment by tsmethurst** [comment]
My interpretation of the logs just posted by @gytisrepecka is that WriteFreely is trying to send an empty url string to the `makeActivityPost` function here: https://github.com/writefreely/writefreely/blob/22de459a72fa21a010ef79944d7cb3a8197fcdb5/activitypub.go#L528C67-L528C71

I think that's because the Actor from GoToSocial that WriteFreely has stored is just the minimal main-key Actor stub that we serve at the main-key endpoint (see here, we're aware this is pretty quirky and will probably change it at some point: https://docs.gotosocial.org/en/latest/federation/federating_with_gotosocial/#quirks). We serve this stub actor without requiring signed GET requests.

The stub actor looks like this:

```json
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    "https://w3id.org/security/v1"
  ],
  "id": "https://goblin.technology/users/tobi",
  "preferredUsername": "tobi",
  "publicKey": {
    "id": "https://goblin.technology/users/tobi/main-key",
    "owner": "https://goblin.technology/users/tobi",
    "publicKeyPem": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAy2/j9+2G7xrQvBtygrj4\naYHl8hTeZecDBnS+6IBjjEt+QWJ3z0Cv9lXSVMZw5i6DVTkVOGZrh2vZDu0BCTEV\n07dyASArE63Qe21WwjNObpkQ7YZbMxUkYjWCDYdqLMifAqElYzIK7xnY0pTWylmC\njm39qxmhk22PpzRkw+zofh9ykqyadmkA2/KrpZgGnjn6MiPqb2DeELV1tzmZ7mAz\n+k7pkkhxvBVqPhCZ104pyd1lc66obONSnIqxugRlrrUZbFv1e6xFsUmMUYrAGQTt\nZr4VeZwuaYj/MvFIeOZrmth/lg3QpYbKZYnJKVePyH+530jRSFerr2unbuGmEQAg\nvQIDAQAB\n-----END PUBLIC KEY-----\n"
  },
  "type": "Person"
}
```

So I would guess that WriteFreely is looking there for an inbox URI, not finding it and falling back to an empty string, hence the above `unsupported protocol scheme` error.

To make this work with GoToSocial, changes will be required so that WriteFreely first fetches whatever is served at the `main-key` address, then uses the `owner` field to make a signed GET request to fetch the actual Actor model that owns the key. This isn't GtS specific; these changes would maintain compatibility with other fedi implementations too -- at worst you'll make two dereferences in a row the first time you see a new Actor.

It it helps, this is also how we do it ourselves, here: https://github.com/superseriousbusiness/gotosocial/blob/b14cd645f9e84c88cbf1d0ea6532e10e571ac1fe/internal/federation/authenticate.go#L104-L236

**Comment by cmw** [comment]
@thebaer is there a plan to address this? I would love to keep following writefreely accounts from my GTS instance.

I can confirm the issue persists in the latest version of GoToSocial as described above. Unfortunately I don't know enough about ActivityPub or Go to submit a PR. I am, however, happy to support with testing, if that helps.

**Comment by CDN18** [comment]
With the help of [tobi's comment](https://github.com/writefreely/writefreely/issues/661#issuecomment-1951367449), I made the corresponding changes to `getActor()` in #1000. My Writefreely instance can now interact properly with GoToSocial.

**Comment by thebaer** [comment]
Appreciate everyone digging into this and @CDN18 submitting a fix! I'll get #1000 reviewed and merged for our next release (v0.15.1).
